libs/@local/harpc/crates/wire/src/request/mod.rs

```diff
@@ -0,0 +1,327 @@
+pub use bytes::Bytes;
+use error_stack::{Result, ResultExt};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+use self::{
+    body::{RequestBody, RequestBodyContext},
+    codec::{DecodeError, EncodeError},
+    header::RequestHeader,
+};
+use crate::codec::{Decode, DecodePure, Encode};
+
+pub mod authorization;
+pub mod begin;
+pub mod body;
+pub mod codec;
+pub mod encoding;
+pub mod flags;
+pub mod frame;
+pub mod header;
+pub mod id;
+pub mod procedure;
+pub mod service;
+
+/// A request message.
+///
+/// A request message is composed of a header and a body and is used to send the actual message over
+/// the wire, agnostic over the transport layer.
+///
+/// This is a binary protocol with minimal overhead with framing support. Messages can be up to
+/// 64KiB in size, larger messages are split into multiple frames.
+///
+/// The transport layer is responsible for ensuring that the message is sent and received in order.
+///
+///
+/// # `Begin` Packet
+///
+/// The layout of a `Begin` packet is as follows:
+///
+/// ```text
+/// 0                   1                   2                   3
+/// 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |  Magic  |P|R. |F|S. |S. |P. |E. |A. |      Authorization      |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |     |P. |                       Payload                       |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///
+/// * Magic (5 bytes)
+/// * Protocol Version (1 byte)
+/// * Request Id (2 bytes)
+/// * Flags (1 byte)
+/// * Service Id (2 bytes)
+/// * Service Version (2 bytes)
+/// * Procedure Id (2 bytes)
+/// * Encoding (2 bytes)
+/// * Accept (2 bytes)
+/// * Authorization (16 bytes)
+/// * Payload Length (2 bytes)
+/// * Payload (27 bytes)
```

Comment:
If the above suggestions would be applied, the payload would be 11 bytes (16 bytes request id, no encoding, 8 bytes payload length), which leads to the question if the `begin` frame should contain a payload at all. This is also questionable as the payload would be misaligned which will have an impact on performance.

Reply:
I see now that me generating the illustration was misleading, the payload can be up to 64KiB in size, for every packet (payload length only denotes the length of the payload, not the whole response) I will reword this.

Reply:
illustration has been updated